chain/network/src/routing_table_actor.rs

```diff
@@ -14,22 +16,355 @@ use crate::ibf::{Ibf, IbfBox};
+            false
+        } else {
+            self.needs_routing_table_recalculation = true;
+            match edge.edge_type() {
+                EdgeType::Added => {
+                    self.raw_graph.add_edge(key.0.clone(), key.1.clone());
+                }
+                EdgeType::Removed => {
+                    self.raw_graph.remove_edge(&key.0, &key.1);
+                }
+            }
+            #[cfg(feature = "protocol_feature_routing_exchange_algorithm")]
+            self.peer_ibf_set.add_edge(&edge.to_simple_edge());
+            self.edges_info.insert(key, edge);
+            true
+        }
+    }
+
+    /// Add several edges to the current view of the network.
+    /// These edges are assumed to have been verified at this point.
+    /// Return list of edges added.
+    ///
+    /// Everytime we remove an edge we store all edges removed at given time to disk.
+    /// If new edge comes comes that is adjacent to a peer that has been previously removed,
+    /// we will try to re-add edges previously removed from disk.
+    pub fn add_verified_edges_to_routing_table(&mut self, mut edges: Vec<Edge>) -> Vec<Edge> {
+        let total = edges.len();
+        edges.retain(|edge| {
+            let key = edge.get_pair();
+
+            self.fetch_edges_for_peer_from_disk(&key.0);
+            self.fetch_edges_for_peer_from_disk(&key.1);
+
+            self.add_verified_edge(edge.clone())
+        });
+
+        // Update metrics after edge update
+        near_metrics::inc_counter_by(&metrics::EDGE_UPDATES, total as u64);
+        near_metrics::set_gauge(&metrics::EDGE_ACTIVE, self.raw_graph.total_active_edges as i64);
+
+        edges
+    }
+
+    /// If peer_id is not in memory check if it is on disk in bring it back on memory.
+    fn fetch_edges_for_peer_from_disk(&mut self, other_peer_id: &PeerId) {
+        if other_peer_id == self.my_peer_id()
+            || self.peer_last_time_reachable.contains_key(other_peer_id)
+        {
+            return;
+        }
+
+        let my_peer_id = self.my_peer_id().clone();
+
+        // Get the "row" (a.k.a nonce) at which we've stored a given peer in the past (when we pruned it).
+        if let Ok(component_nonce) = self.component_nonce_from_peer(other_peer_id.clone()) {
+            let mut update = self.store.store_update();
+
+            // Load all edges that were persisted in database in the cell - and add them to the current graph.
+            if let Ok(edges) = self.get_and_remove_component_edges(component_nonce, &mut update) {
+                for edge in edges {
+                    for &peer_id in vec![&edge.peer0, &edge.peer1].iter() {
+                        if peer_id == &my_peer_id
+                            || self.peer_last_time_reachable.contains_key(peer_id)
+                        {
+                            continue;
+                        }
+
+                        // We remove mapping from `peer_id` to component_nonce else if
```

Comment:
this comment is a little bit confusing.

I'd write something like this:

(other_peer_id, peer_id) is the edge that belongs to the component that we've recovered from disk. Now we're checking what happened to peer_id. If it was also the part of the same component - great, let's mark it as reachable and remove the reference in DB.
If it was a part of another component - let's not touch it.

Reply:
Ok, I updated the code. I also added warnings, because other cases shouldn't happen unless node crashes in middle of update.